README.md
chi-elections
=============

chi-elections is a Python package for loading and parsing election results from the [Chicago Board of Elections](http://www.chicagoelections.com/).

Summary Results
---------------

The Board of Elections provides election-night results at a racewide level.  The file lives at

http://www.chicagoelections.com/results/ap/summary.txt

before election day, for testing.

It lives at

http://www.chicagoelections.com/ap/summary.txt

on election night.

### Text layout

From http://www.chicagoelections.com/results/ap/text_layout.txt:

Summary Export File Format      Length  Column Position
Contest Code                    4       1-4
Candidate Number                3       5-7
Num. Eligible Precincts         4       8-11
Votes                           7       12-18
Num. Completed precincts        4       19-22
Party Abbreviation              3       23-25
Political Subdivision Abbrev    7       26-32
Contest name                    56      33-88
Candidate Name                  38      89-126
Political subdivision name      25      127-151
Vote For                        3       152-154

### Gotchas

Prior to election night, the test file will include all fields.  At some point on election night, the file will only contain the numeric values in the first 22 columns.

This means that you need to:

* Make sure you save candidate names in some way, like a database, before election night
* Make sure you store ballot order (Candidate Number in the text layout above) with the candidate.  You'll need to use this, in combination with Contest Code, to look up the cached candidates.

At some point at the end of election night, the results file will no longer be available at http://www.chicagoelections.com/ap/summary.txt and will be available at http://www.chicagoelections.com/results/ap/summary.txt
.  However, it will not be updated. You'll need to scrape, enter or load results in some other way if you need updates after election night.


### Results client

To access the results:

    from chi_elections import SummaryClient

    client = SummaryClient()
    client.fetch()
    mayor = next(r for r in client.races if r.name == "Mayor")
    self.assertEqual(len(mayor.candidates), 5)

    rahm = next(c for c in mayor.candidates
                if c.full_name == "RAHM EMANUEL")
    print(rahm.vote_total)

If you want to specify an alternate url, for example the test URL, pass it to the constructor of `SummaryClient`:

    client = SummaryClient(url='http://www.chicagoelections.com/results/ap/summary.txt')


Precinct Results
----------------

After election night, precinct-level results are published to http://www.chicagoelections.com/en/election3.asp.  The results are HTML files, so we have to scrape the results from HTML tables.

### Results client

To access the results:

    from chi_elections import PrecinctClient

    client = PrecinctClient()
    client.fetch_elections()
    election = next(e for e in client.elections
                    if e.name == "2015 Municipal Runoffs - 4/7/15")
    election.fetch_races()
    race = next(r for r in client.races
                if r.name == "Mayor")
    race.fetch_results()

### Command Line Interface

To download a CSV version of the summary file, run:

    chi_elections summary > results.csv

To hit the test file:

    chi_elections summary --test > results.csv

To download precinct results, available the day after:

    chi_elections precincts --race "Delegate, National Convention 4th DEM" 5

In this example the `5` is the election code, that can be found in the URL when you visit a page like http://www.chicagoelections.com/en/wdlevel3.asp?elec_code=5


CHANGELOG.md
CHANGELOG
=========

0.2
---

* Handle utf-8 characters
* Replace two single quotes in candidate names with a single double quote
* Add command line interface to output CSV of results


tests/test_summary.py
# -*- coding=utf-8 -*-
import os.path
from unittest import TestCase

import responses

from chi_elections.summary import (FixedWidthField, ResultParser, SummaryClient,
        SummaryParser)

TEST_DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)),
    'data')
SUMMARY_TEST_FILENAME = os.path.join(TEST_DATA_DIR, 'results', 'ap',
    'summary.txt')

class ParserTestCase(TestCase):

    def setUp(self):
        self.parser = SummaryParser()

    def test_parse(self):
        with open(SUMMARY_TEST_FILENAME, 'r') as f:
            self.parser.parse(f.read())
            self.assertEqual(len(self.parser.races), 98)

            mayor = next(r for r in self.parser.races if r.name == "Mayor")
            self.assertEqual(len(mayor.candidates), 5)

            rahm = next(c for c in mayor.candidates
                        if c.full_name == "RAHM EMANUEL")
            self.assertEqual(rahm.vote_total, 0)

           
class FixedWidthFieldTestCase(TestCase):
    def test_parse(self):
        line = "0010001206900000000000NON       Mayor                                                   RAHM EMANUEL                          City Of Chicago          001"
        field = FixedWidthField(0, 3, transform=int)
        parsed = field.parse(line)
        self.assertEqual(parsed, 1)
        field = FixedWidthField(22, 3)
        parsed = field.parse(line)
        self.assertEqual(parsed, "NON")
        field = FixedWidthField(32, 56)
        parsed = field.parse(line)
        self.assertEqual(parsed, "Mayor")



class ResultParserTestCase(TestCase):
    def test_parse_line(self):
        parser = ResultParser()
        line = "0010001206900000000000NON       Mayor                                                   RAHM EMANUEL                          City Of Chicago          001"
        result = parser.parse_line(line)
        self.assertEqual(result['contest_code'], 10)
        self.assertEqual(result['candidate_number'], 1)
        self.assertEqual(result['precincts_total'], 2069)
        self.assertEqual(result['vote_total'], 0)
        self.assertEqual(result['precincts_reporting'], 0)
        self.assertEqual(result['party'], "NON")
        self.assertEqual(result['race_name'], "Mayor")
        self.assertEqual(result['candidate_name'], "RAHM EMANUEL")
        self.assertEqual(result['reporting_unit_name'], "City Of Chicago")
        self.assertEqual(result['vote_for'], 1)

    def test_parse_line_no_text(self):
        parser = ResultParser()
        line = "0010001206900000000000"
        result = parser.parse_line(line)
        self.assertEqual(result['contest_code'], 10)
        self.assertEqual(result['candidate_number'], 1)
        self.assertEqual(result['precincts_total'], 2069)
        self.assertEqual(result['vote_total'], 0)
        self.assertEqual(result['precincts_reporting'], 0)
        self.assertEqual(result['party'], "")
        self.assertEqual(result['race_name'], "")
        self.assertEqual(result['candidate_name'], "")
        self.assertEqual(result['reporting_unit_name'], "")
        self.assertEqual(result['vote_for'], None)

    def test_parse_line_utf8(self):
        parser = ResultParser()
        line = "0023012034800000000000DEM       Delegate, National Convention 4th DEM                   Álvaro R. Obregón (Sanders)           4th Congressional Distric005"
        result = parser.parse_line(line)
        self.assertEqual(result['contest_code'], 23)
        self.assertEqual(result['candidate_number'], 12)
        self.assertEqual(result['precincts_total'], 348)
        self.assertEqual(result['vote_total'], 0)
        self.assertEqual(result['precincts_reporting'], 0)
        self.assertEqual(result['party'], "DEM")
        self.assertEqual(result['race_name'], "Delegate, National Convention 4th DEM")
        self.assertEqual(result['candidate_name'], u"Álvaro R. Obregón (Sanders)")
        self.assertEqual(result['reporting_unit_name'], "4th Congressional Distric")
        self.assertEqual(result['vote_for'], 5)


class SummaryClientTestCase(TestCase):
    @responses.activate
    def test_fetch(self):
        client = SummaryClient()
        with open(SUMMARY_TEST_FILENAME) as f:
            response_body = f.read()
            responses.add(responses.GET, client.get_url(), body=response_body,
                content_type='text/plain')    
            client.fetch() 
            self.assertEqual(len(client.races), 98)

            mayor = next(r for r in client.races if r.name == "Mayor")
            self.assertEqual(len(mayor.candidates), 5)

            rahm = next(c for c in mayor.candidates
                        if c.full_name == "RAHM EMANUEL")
            self.assertEqual(rahm.vote_total, 0)


tests/test_transforms.py
import unittest

from chi_elections.transforms import replace_single_quotes

class TestTransforms(unittest.TestCase):
    def test_replace_single_quotes(self):
        self.assertEqual(replace_single_quotes("Roque ''Rocky'' De La Fuente"),
            'Roque "Rocky" De La Fuente')


chi_elections/cli.py
import codecs
import sys

if sys.version_info < 3:
    from backports import csv
else:
    import csv

import click

from .constants import SUMMARY_URL, TEST_SUMMARY_URL
from .precincts import Election
from .summary import SummaryClient, SummaryParser

if sys.version_info < 3:
    # Wrap sys.stdout into a StreamWriter to allow writing unicode.
    # See https://wiki.python.org/moin/PrintFails
    sys.stdout = codecs.getwriter('utf-8')(sys.stdout)

@click.group()
def main():
    pass

@click.command()
@click.option('-f', '--file', type=click.File())
@click.option('--test/--no-test', default=False)
def summary(file, test):
    if test:
        url = TEST_SUMMARY_URL
    else:
        url = SUMMARY_URL

    if file:
        parser = SummaryParser()
        parser.parse(file.read())
        races = parser.races
    else:
        client = SummaryClient(url=url)
        client.fetch()
        races = client.races

    fieldnames = [
       'contest_code',
       'race_name',
       'precincts_total',
       'precincts_reporting',
       'vote_for',
       'candidate_number',
       'full_name',
       'party',
       'vote_total',
    ]
    writer = csv.DictWriter(sys.stdout, fieldnames=fieldnames)
    writer.writeheader()
    for race in races:
        race_attrs = race.serialize()
        for candidate_result in race.candidates:
            row = dict(**race_attrs)
            row.update(candidate_result.serialize())
            writer.writerow(row)

main.add_command(summary)


@click.command()
@click.argument('elections', nargs=-1)
@click.option('--race', '-r', default=None, multiple=True)
def precincts(elections, race):
    fieldnames = [
       'race_name',
       'race_number',
       'candidate',
       'ward',
       'precinct',
       'votes',
    ]
    writer = csv.DictWriter(sys.stdout, fieldnames=fieldnames)
    writer.writeheader()
    for election_id in elections:
        election = Election(elec_code=election_id)
        if len(race):
            races_set = set()
            for rid in race:
                try:
                    race_identifier = int(rid)
                except ValueError:
                    race_identifier = rid

                races_set.add(race_identifier)

            races = [r for r in election.races
                     if r.number in races_set or r.name in races_set]

        else:
            races = election.races

        for race in races:
            for result in race.results:
                try:
                    writer.writerow(result.serialize())
                except UnicodeDecodeError:
                    print(result.serialize())
                    raise

main.add_command(precincts)


chi_elections/precincts.py
"""
Parse tabular precinct-level results.
"""
from collections import OrderedDict

from lxml import html
import requests
from six.moves.urllib.parse import urlencode, urlparse, parse_qs
from six import text_type

from chi_elections.transforms import replace_single_quotes

class BaseParser(object):
    @classmethod
    def clean_cell(cls, s):
        return text_type(s).strip()

    def get_row_data(self, tr):
        return [self.clean_cell(td.text_content()) for td in tr.xpath('td')]

    @classmethod
    def clean_candidate_name(cls, s):
        return replace_single_quotes(s)

    def parse_candidates(self, row):
        candidates = {}
        for i, val in enumerate(row):
            if val in (self.reporting_unit_column_name, "%"):
                continue

            candidates[i] = self.clean_candidate_name(val)

        return candidates

    def parse_result_row(self, row, candidate_lookup):
        results = []
        result = None
        for i, val in enumerate(row):
            if i == 0:
                reporting_unit_id = val
                continue

            try:
                candidate = candidate_lookup[i]
                if result is not None:
                    # If a previous result has been built, add it to our list
                    results.append(result)
                result = {
                    'reporting_unit_id': reporting_unit_id,
                    'candidate': candidate,
                    'votes': int(val),
                }
            except KeyError:
                # No candidate at this column index.
                # Assume it's the percentage column
                result['percent'] = float(val.strip('%'))

        if result is not None:
            results.append(result)

        return results

    def parse(self, html_string):
        results = []

        rows = html.fromstring(html_string).xpath('//table[1]/tr')
        candidate_lookup = None
        for tr_idx, tr in enumerate(rows):
            row = self.get_row_data(tr)
            if len(row) < 2:
                # This is a blank row, so skip
                continue

            if row[0] == self.reporting_unit_column_name:
                # This is the header row
                if candidate_lookup is None:
                    candidate_lookup = self.parse_candidates(row)
                continue

            row_results = self.parse_result_row(row, candidate_lookup)
            results.extend(row_results)

        return results


class WardParser(BaseParser):
    reporting_unit_column_name = 'Ward'


class PrecinctParser(BaseParser):
    reporting_unit_column_name = 'Pct'


class ReportingUnit(object):
    def __init__(self, number):
        self.number = number
        self._results = []

    def add_result(self, result):
        self._results.append(result)

    @property
    def results(self):
        return self._results


class Ward(ReportingUnit):
    level = 'ward'

    def __init__(self, number):
        super(Ward, self).__init__(number)
        self._precincts = {}

    def get_or_create_precinct(self, number):
        try:
            return self._precincts[number]
        except KeyError:
            self._precincts[number] = Precinct(number, self)
            return self._precincts[number]

    def __str__(self):
        return "{:02d}".format(self.number)

    def __repr__(self):
        return "Ward({})".format(self.__str__())


class Precinct(ReportingUnit):
    level = 'precinct'

    def __init__(self, number, ward):
        super(Precinct, self).__init__(number)
        self.ward = ward

    def __str__(self):
        return "{:02d}{:03d}".format(int(self.ward.number),
            int(self.number))

    def __repr__(self):
        return "Precinct({})".format(self.__str__())


class Candidate(object):
    def __init__(self, name):
        self.name = name

    def __str__(self):
        return self.name

    def __repr__(self):
        return "Candidate({})".format(self.name)


class Election(object):
    def __init__(self, elec_code, name=None, client=None):
        self.elec_code = elec_code
        self.name = name

        if client is None:
            self.client = PrecinctClient()

        self._races_by_number = None
        self._races_by_name = None
        self._races = None

    @property
    def races(self):
        if self._races is None:
            self.fetch_races()

        return self._races

    def add_race(self, race):
        if race.number is not None:
            self._races_by_number[race.number] = race

        if race.name is not None:
            self._races_by_name[race.name] = race

        self._races.append(race)

    def get_race_by_number(self, race_number):
        return self._races_by_number[race_number]

    def get_race_by_name(self, race_name):
        return self._races_by_name[race_name]

    def fetch_races(self):
        race_html = self.client.fetch_election_html(self.elec_code)
        option_els = html.fromstring(race_html).xpath(
            "//select[@name='D3']/option")

        self._races_by_number = {}
        self._races_by_name = {}
        self._races = []

        for option_el in option_els:
            race_name = option_el.get('value')
            if not race_name:
                continue
            race = Race(self, name=race_name)


class Race(object):
    def __init__(self, election, name=None, number=None):
        self.number = number
        self.name = name
        self.election = election
        self.client = election.client

        election.add_race(self)

        self._results = None
        self._wards = None

    def __str__(self):
        if self.name and self.number:
            return "{} ({})".format(self.name, self.number)
        elif self.name:
            return self.name
        elif self.number:
            return self.number
        else:
            return ""

    @property
    def results(self):
        if self._results is None:
            self._results = self.fetch_results()

        return self._results

    @property
    def wards(self):
        if self._wards is None:
            self.fetch_wards()

        return self._wards.values()

    def fetch_wards(self):
        ward_results_html = self.client.fetch_ward_results_html(
            elec_code=self.election.elec_code,
            race_name=self.name
        )

        if self.number is None:
            first_ward_url = html.fromstring(ward_results_html).xpath('//a[1]')[0].get('href')
            # TODO: Factor this out
            query_string = urlparse(first_ward_url).query
            query_string_parsed = parse_qs(query_string)
            self.number = int(query_string_parsed['race_number'][0])

        parser = WardParser()
        results_attrs = parser.parse(ward_results_html)
        self._wards = {}

        for result_attrs in results_attrs:
            if result_attrs['reporting_unit_id'] == "Total":
                continue

            ward_number = int(result_attrs['reporting_unit_id'])
            self._wards.setdefault(ward_number, Ward(
                number=ward_number,
            ))

    def fetch_results(self):
        results = []
        for ward in self.wards:
            ward_results =  self.client.fetch_precinct_results(
                elec_code=self.election.elec_code,
                race=self,
                ward_num=ward.number)
            results.extend(ward_results)

        return results


class Result(object):
    def __init__(self, candidate, votes, reporting_unit, percent=None, race=None):
        self.candidate = candidate
        self.votes = votes
        self.reporting_unit = reporting_unit
        self.percent = percent
        self.race = race

        self.reporting_unit.add_result(self)

    def __str__(self):
        return "{} - {} ({})".format(self.candidate, self.votes,
            self.reporting_unit)

    def __repr__(self):
        return "Result(candidate={}, reporting_unit={}, votes={})".format(
            repr(self.candidate), repr(self.reporting_unit), self.votes
        )

    @property
    def ward_number(self):
        if self.reporting_unit.level == "ward":
            return self.reporting_unit.number
        elif self.reporting_unit.level == "precinct":
            return self.reporting_unit.ward.number
        else:
            return None

    @property
    def precinct_number(self):
        if self.reporting_unit.level == "precinct":
            return self.reporting_unit.number
        else:
            return None

    def serialize(self):
        return OrderedDict((
            ('race_name', self.race.name),
            ('race_number', self.race.number),
            ('candidate', self.candidate.name),
            ('ward', self.ward_number),
            ('precinct', self.precinct_number),
            ('votes', self.votes),
        ))


class PrecinctClient(object):
    DEFAULT_PRECINCT_URL = 'http://www.chicagoelections.com/en/pctlevel3.asp'
    DEFAULT_ELECTION_URL = 'http://www.chicagoelections.com/en/wdlevel3.asp'

    def __init__(self, election_url=None, precinct_url=None):
        if election_url is None:
            election_url = self.DEFAULT_ELECTION_URL

        self._election_url = election_url

        if precinct_url is None:
            precinct_url = self.DEFAULT_PRECINCT_URL

        self._precinct_url = precinct_url

        self._parser = PrecinctParser()
        self._wards = {}
        self._candidates_by_name = {}

    def get_election_url(self, elec_code):
        url = self._election_url
        query_params = {
            'elec_code': elec_code,
        }
        qs = urlencode(query_params)
        return url + '?' + qs

    def fetch_election_html(self, elec_code):
        url = self.get_election_url(elec_code)
        return requests.get(url).text

    def fetch_ward_results_html(self, elec_code, race_name):
        url = self.get_election_url(elec_code)
        return requests.post(url, data={
                'VTI-GROUP': 0,
                'flag': 1,
                'B1': 'View The Results',
                'D3': race_name,
            }).text

    def get_precinct_result_url(self, elec_code, race_number, ward):
        url = self._precinct_url
        query_params = {
            'elec_code': elec_code,
            'race_number': race_number,
            'Ward': ward,
        }
        qs = urlencode(query_params)
        return url + '?' + qs

    def fetch_precinct_results_html(self, elec_code, race_number, ward):
        url = self.get_precinct_result_url(elec_code, race_number, ward)
        return requests.get(url).text

    def get_or_create_ward(self, ward_num):
        try:
            return self._wards[ward_num]
        except KeyError:
            self._wards[ward_num] = Ward(ward_num)
            return self._wards[ward_num]

    def get_or_create_candidate_by_name(self, name):
        try:
            return self._candidates_by_name[name]
        except KeyError:
            self._candidates_by_name[name] = Candidate(name)
            return self._candidates_by_name[name]

    def create_ward_result(self, result_dict, ward):
        candidate = self.get_or_create_candidate_by_name(result_dict['candidate'])
        return Result(candidate, result_dict['votes'],
            percent=result_dict['percent'], reporting_unit=ward)

    def create_result(self, result_dict, race, ward_num):
        ward = self.get_or_create_ward(ward_num)
        if result_dict['reporting_unit_id'] == 'Total':
            reporting_unit = ward
        else:
            reporting_unit = ward.get_or_create_precinct(result_dict['reporting_unit_id'])

        candidate = self.get_or_create_candidate_by_name(result_dict['candidate'])

        if isinstance(race, Race):
            race_model = race
        else:
            race_model = Race(number=race)

        return Result(
            candidate=candidate,
            votes=result_dict['votes'],
            reporting_unit=reporting_unit,
            percent=result_dict.get('percent', None),
            race=race_model
        )

    def fetch_precinct_results(self, elec_code, race, ward_num):
        if isinstance(race, Race):
            race_number = race.number
        else:
            race_number = race

        html_string = self.fetch_precinct_results_html(elec_code, race_number,
            ward_num)
        results = self._parser.parse(html_string)
        return [self.create_result(rd, race, ward_num) for rd in results]


chi_elections/summary.py
"""
Parse fixed-width summary file.

This file lives at

http://www.chicagoelections.com/results/ap/

before election day, for testing.

It lives at

www.chicagoelections.com/ap/

on election night.

This file provides racewide results.

"""
from collections import OrderedDict

import requests

from .constants import SUMMARY_URL
from .transforms import replace_single_quotes

class FixedWidthField(object):
    def __init__(self, index, length, transform=None):
        self.index = index
        self.length = length
        self.transform = transform
        self.name = None

    def parse(self, s):
        try:
            s_decoded = s.decode('utf-8')
        except UnicodeEncodeError:
            s_decoded = s

        val = s_decoded[self.index:self.index + self.length]
        val = val.strip()
        if self.transform is None:
            return val
        else:
            try:
                return self.transform(val)
            except ValueError:
                return None


class FixedWidthParserMeta(type):
    def __new__(cls, name, parents, dct):
        dct['_fields'] = []
        for k, v in dct.items():
            if isinstance(v, FixedWidthField):
                v.name = k
                dct['_fields'].append(v)
                del dct[k]

        new_cls = super(FixedWidthParserMeta, cls).__new__(cls, name, parents, dct)
        return new_cls


class FixedWidthParser(object):
    __metaclass__ = FixedWidthParserMeta

    def parse_line(self, line):
        attrs = {}
        for field in self._fields:
            attrs[field.name] = field.parse(line)

        return attrs 


class ResultParser(FixedWidthParser):
    # Summary Export File Format           Length    Column Position
    # Contest Code                         4         1-4
    # Candidate Number                     3         5-7
    # # of Eligible Precincts              4         8-11
    # Votes                                7         12-18
    # # Completed precincts                4         19-22
    # Party Abbreviation                   3         23-25
    # Political Subdivision Abbreviation   7         26-32
    # Contest name                         56        33-88
    # Candidate Name                       38        89-126
    # Political subdivision name           25        127-151
    # Vote For                             3         152-154
    contest_code = FixedWidthField(0, 4, transform=int)
    candidate_number = FixedWidthField(4, 3, transform=int)
    precincts_total = FixedWidthField(7, 4, transform=int)
    vote_total = FixedWidthField(11, 7, transform=int)
    precincts_reporting = FixedWidthField(18, 4, transform=int)
    party = FixedWidthField(22, 3)
    reporting_unit_name = FixedWidthField(25, 7)
    race_name = FixedWidthField(32, 56)
    candidate_name = FixedWidthField(88, 38, transform=replace_single_quotes)
    reporting_unit_name = FixedWidthField(126, 25)
    vote_for = FixedWidthField(151, 3, transform=int)


class Result(object):
    def __init__(self, candidate_number, full_name, party, race, vote_total,
            reporting_unit_name):
        self.candidate_number = candidate_number
        self.full_name = full_name
        self.party = party
        self.race = race
        self.vote_total = vote_total

    def __str__(self):
        return "{}: {}d".format(self.name, self.vote_total)

    def serialize(self):
        return OrderedDict((
            ('candidate_number', self.candidate_number),
            ('full_name', self.full_name),
            ('party', self.party),
            ('vote_total',self.vote_total),
        ))


class Race(object):
    def __init__(self, contest_code, name, precincts_total=0,
            precincts_reporting=0, vote_for=1):
        self.contest_code = contest_code
        self.name = name
        self.candidates = []
        self.precincts_total = precincts_total
        self.precincts_reporting = precincts_reporting
        self.vote_for = vote_for

    def serialize(self):
        return OrderedDict((
            ('contest_code', self.contest_code),
            ('race_name', self.name),
            ('precincts_total', self.precincts_total),
            ('precincts_reporting', self.precincts_reporting),
            ('vote_for', self.vote_for),
        ))

    def __str__(self):
        return self.name


class SummaryParser(object):
    def __init__(self):
        self._result_parser = ResultParser()

    def parse(self, s):
        self.races = []
        self._race_lookup = {}

        for line in s.splitlines(True):
            parsed = self._result_parser.parse_line(line)
            race = self.get_or_create_race(parsed)
            result = Result(
                candidate_number=parsed['candidate_number'],    
                vote_total=parsed['vote_total'],
                party=parsed['party'],
                race=race,
                full_name=parsed['candidate_name'],
                reporting_unit_name=parsed['reporting_unit_name'],
            )
            race.candidates.append(result)
    
    def get_or_create_race(self, attrs):
        try:
            race = self._race_lookup[attrs['contest_code']]
        except KeyError:
            race = Race(
                contest_code=attrs['contest_code'],
                name=attrs['race_name'],
                precincts_total=attrs['precincts_total'],
                precincts_reporting=attrs['precincts_reporting'],
                vote_for=attrs['vote_for'],
            )
            self._race_lookup[attrs['contest_code']] = race
            self.races.append(race)
        
        return race


class SummaryClient(object):
    DEFAULT_URL = SUMMARY_URL 

    def __init__(self, url=None):
        if url is None:
            url = self.DEFAULT_URL
        self._url = url

        self._parser = SummaryParser()

    def get_url(self):
        return self._url

    def fetch(self):
        url = self.get_url()
        r = requests.get(url)
        self._parser.parse(r.text)

    @property
    def races(self):
        return self._parser.races
